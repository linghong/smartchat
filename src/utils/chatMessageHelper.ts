/**
 * Extracts the message content by removing the <meta> tags and subject title.
 * @param message - The full message including the subject title in <meta> tags.
 * @returns The message content without the <meta> tags.
 */
export const extractMessageContent = (message: string): string => {
  return message.replace(/\{\{\{.*\}\}\}$/, '').trim()
}

/**
 * Extracts the subject title from the message.
 * @param message - The full message including the subject title in <meta> tags.
 * @returns The subject title if found, otherwise an empty string.
 */
export const extractSubjectTitle = (message: string): string => {
  const match = message.match(/\{\{\{(.*)\}\}\}$/)
  return match ? match[1].trim() : ''
}