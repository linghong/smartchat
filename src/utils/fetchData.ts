const API_URL = process.env.NEXT_PUBLIC_API_URL

export const fetchData = async (endpoint: string) => {
  try {
    const response = await fetch(`${API_URL}/api/${endpoint}`);

    if (!response.ok) {
      throw new Error(`Error fetching ${endpoint}: ${response.statusText}`);
    }

    const data = await response.json();
    return data;

  } catch (error) {
    console.error(`Error fetching ${endpoint}:`, error);
    return {
      props: {
        error: "Failed to fetch namespaces."
      }
    };
  }
};