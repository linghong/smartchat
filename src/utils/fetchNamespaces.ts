import { getNamespaces } from '@/src/services/pineconeClient'
import { PINECONE_INDEX_NAME } from '@/config/env'

export const fetchNamespaces = async () => {
 
  try {
    if (!PINECONE_INDEX_NAME) {
      console.error('Missing Pinecone index name')
      return []
    }
    const namespaces = await getNamespaces(PINECONE_INDEX_NAME)
    
    return namespaces
    
  } catch (error) {
    console.error(`Error fetching namespaces:`, error)
    return []
  }
}
