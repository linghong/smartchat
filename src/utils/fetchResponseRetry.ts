const MAX_RETRIES = 3;
const RETRY_DELAY = 1000; // 1 second

export const delay = (ms: number) =>
  new Promise(resolve => setTimeout(resolve, ms));

export const handleErrors = async (error: any, retryCount: number) => {
  if (retryCount >= MAX_RETRIES) {
    throw error;
  }

  await delay(RETRY_DELAY * (retryCount + 1));
};
