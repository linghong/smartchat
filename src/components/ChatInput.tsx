import React, {
  useState,
  useEffect,
  useCallback,
  useRef,
  ChangeEvent,
  FormEvent,
  Dispatch,
  SetStateAction
} from 'react';
import DOMPurify from 'isomorphic-dompurify';
import { RiScreenshot2Fill } from 'react-icons/ri';

import { FileData } from '@/src/types/chat';
import { OptionType } from '@/src/types/common';

import ArrowButton from '@/src/components/ArrowButton';
import ButtonWithTooltip from '@/src/components/ButtonWithTooltip';
import FileUploadIcon from '@/src/components/FileUploadIcon';

import FileListWithModal from '@/src/components/FileListWithModal';
import Notification from '@/src/components/Notification';

import {
  fileToDataURLBase64,
  fileToArrayBufferBase64
} from '@/src/utils/fileFetchAndConversion';
import { isSupportedImage } from '@/src/utils/mediaValidationHelper';

interface ChatInputProps {
  onSubmit: (question: string, fileSrc: FileData[]) => void;
  isVisionModel: boolean;
  selectedModel: OptionType;
  isConfigPanelVisible: boolean;
  setIsConfigPanelVisible: Dispatch<SetStateAction<boolean>>;
}

const ChatInput: React.FC<ChatInputProps> = ({
  onSubmit,
  isVisionModel,
  selectedModel,
  isConfigPanelVisible,
  setIsConfigPanelVisible
}) => {
  const textAreaRef = useRef<HTMLTextAreaElement>(null);
  const [userInput, setUserInput] = useState('');
  const [rows, setRows] = useState<number>(1);
  const [fileSrc, setFileSrc] = useState<FileData[]>([]);
  const [fileError, setFileError] = useState<string[]>([]);

  const handleSubmit = useCallback(
    async (e: FormEvent<HTMLFormElement> | KeyboardEvent) => {
      e.preventDefault();
      const question: string = userInput.trim();

      if (question.length === 0 && fileSrc.length === 0) return;

      onSubmit(question, fileSrc);

      setUserInput('');
      setFileSrc([]);

      // Reset the textarea rows to initial state
      setRows(1);

      // Reset the textarea height
      if (textAreaRef.current) {
        textAreaRef.current.style.height = 'auto';
      }
    },
    [fileSrc, userInput, onSubmit]
  );

  const handleImageFile = async (fileData: FileData) => {
    if (!isVisionModel) {
      setFileError([...fileError, 'The model only supports text message.']);
      return;
    }
    try {
      const imageVadiationError = isSupportedImage(
        selectedModel?.value || '',
        fileData
      );
      if (imageVadiationError.length !== 0) {
        setFileError([...fileError, ...imageVadiationError]);
        return;
      }

      setFileSrc([...fileSrc, fileData]);
    } catch {
      setFileError([...fileError, 'Failed to read the image file']);
      throw new Error('Failed to read the image file.');
    }
  };

  const handleFileUpload = async (file: File) => {
    if (!file) return;

    const base64Content = await fileToDataURLBase64(file);
    if (!base64Content) {
      setFileError([...fileError, 'Error happened when extracting file data']);
      return;
    }
    const fileType = file.type
      ? file.type
      : base64Content.slice(5).split(';')[0];

    const fileData: FileData = {
      base64Content: DOMPurify.sanitize(base64Content),
      type: fileType,
      size: file.size,
      name: file.name
    };

    if (fileType.startsWith('image/')) {
      await handleImageFile(fileData);
    } else {
      setFileSrc([...fileSrc, fileData]);
    }
  };

  const handleScreenCapture = async () => {
    try {
      const response = await fetch('/api/tools/screenshot', { method: 'POST' });
      const { message, base64Image, type, size, name } = await response.json();

      if (!response.ok) {
        alert(message);
        return;
      }
      const newImage: FileData = {
        base64Content: base64Image,
        type,
        size,
        name
      };

      const imageVadiationError = isSupportedImage(
        selectedModel?.value || '',
        newImage
      );
      if (imageVadiationError.length !== 0) {
        setFileError([...fileError, ...imageVadiationError]);
        return;
      }

      setFileSrc([...fileSrc, newImage]);
    } catch (error) {
      console.error('Error:', error);
      alert('An error occurred while capturing the screen');
    }
  };

  const handleFileDelete = (id: number) => {
    setFileSrc(prevFiles => prevFiles.filter((_, index) => index !== id));
  };

  const handleInputChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    const newValue = e.target.value;

    setUserInput(DOMPurify.sanitize(newValue));

    // Close config panel when user starts typing
    if (newValue.length >= 1 && isConfigPanelVisible) {
      setIsConfigPanelVisible(false);
    }

    //when copying a paragraph
    const newRows = e.target.value.match(/\n/g)?.length ?? 0;
    setRows(Math.min(newRows + 1, 8));

    // automatically adjust the height of the textarea as the user types
    if (textAreaRef.current) {
      textAreaRef.current.style.height = 'auto';
      textAreaRef.current.style.height = `${textAreaRef.current.scrollHeight}px`;
    }
  };

  // useEffect for Shift + Enter
  useEffect(() => {
    const handleShiftEnter = (e: KeyboardEvent) => {
      if (e.key !== 'Enter') return;
      e.preventDefault();

      if (e.shiftKey) {
        const start = textAreaRef.current?.selectionStart;
        const end = textAreaRef.current?.selectionEnd;

        if (start !== undefined && end !== undefined && textAreaRef.current) {
          const value = textAreaRef.current.value;
          const newValue =
            value.substring(0, start) + '\n' + value.substring(end);
          setUserInput(newValue);
          setRows(rows => rows + 1);

          // Update cursor position
          setTimeout(() => {
            if (textAreaRef.current) {
              textAreaRef.current.selectionStart =
                textAreaRef.current.selectionEnd = start + 1;
            }
          }, 0);
        }
      } else {
        handleSubmit(e);
      }
    };

    const currentTextArea = textAreaRef.current;
    if (currentTextArea) {
      currentTextArea.addEventListener('keydown', handleShiftEnter);
    }

    return () => {
      if (currentTextArea) {
        currentTextArea.removeEventListener('keydown', handleShiftEnter);
      }
    };
  }, [handleSubmit]);

  return (
    <div className="flex flex-col w-full justify-around flex-shrink-0 items-center  my-1">
      {fileSrc.length > 0 && (
        <FileListWithModal
          fileSrc={fileSrc}
          handleFileDelete={handleFileDelete}
          isDeleteIconShow={true}
        />
      )}
      <div className="flex w-full justify-around items-center border-2 border-indigo-300 bg-indigo-200 bg-opacity-30 rounded-lg">
        <div className="flex w-3/12 ms:w-2/12  sm:w-1/12 items-center justify-around md:mx-1">
          <ButtonWithTooltip
            icon={<RiScreenshot2Fill size={30} />}
            onClick={handleScreenCapture}
            ariaLabel="Capture Screenshot"
            tooltipText="Capture Screenshot"
            isDisabled={!isVisionModel}
            tooltipDisabledText={`${selectedModel?.value} does not have vision feature`}
          />
          <ButtonWithTooltip
            icon={
              <FileUploadIcon
                onFileUpload={handleFileUpload}
                isDisabled={!isVisionModel}
              />
            }
            onClick={() => {}}
            ariaLabel="Upload File"
            tooltipText="Upload File"
            isDisabled={!isVisionModel}
            tooltipDisabledText={`${selectedModel?.value} does not have vision feature`}
          />
        </div>
        <form
          onSubmit={handleSubmit}
          className="flex items-center w-8/12 ms:w-9/12 xs:w-10/12 xl:w-11/12 px-2 py-1 rounded-lg hover:bg-indigo-100 focus:outline-none focus:ring focus:ring-stone-300 focus:ring-offset-red"
        >
          <textarea
            ref={textAreaRef}
            autoFocus={false}
            spellCheck="true"
            rows={rows}
            id="userInput"
            name="userInput"
            className="w-full max-h-96 placeholder-gray-400 overflow-y-auto focus: p-3 focus:ring-stone-100 focus:outline-none"
            placeholder="Click to send. Shift + Enter for a new line."
            value={userInput}
            onChange={handleInputChange}
            aria-label="Enter your message here"
          />
          <ArrowButton
            disabled={userInput === '' && fileSrc.length === 0}
            aria-label="Send"
          />
        </form>
      </div>
      {fileError &&
        fileError.map((err, i) => (
          <Notification key={i} type="error" message={err} />
        ))}
    </div>
  );
};

export default ChatInput;
