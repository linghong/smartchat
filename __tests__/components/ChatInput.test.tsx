import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import ChatInput from '@/src/components/ChatInput';
import { sanitizeWithPreserveCode } from '@/src/utils/guardrail';

jest.mock('@/src/utils/mediaValidationHelper', () => ({
  isSupportedImage: jest.fn().mockReturnValue([])
}));

jest.mock('@/src/utils/guardrail', () => ({
  sanitizeWithPreserveCode: jest.fn(input => input)
}));

jest.mock('@/src/utils/fileFetchAndConversion', () => ({
  fileToDataURLBase64: jest.fn().mockResolvedValue('mocked-base64-content')
}));

describe('ChatInput', () => {
  const mockOnSubmit = jest.fn();
  const mockSetIsConfigPanelVisible = jest.fn();

  const defaultProps = {
    onSubmit: mockOnSubmit,
    isVisionModel: true,
    selectedModel: {
      value: 'ai-model',
      label: 'AI model',
      contextWindow: 8192
    },
    isConfigPanelVisible: false,
    setIsConfigPanelVisible: mockSetIsConfigPanelVisible
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    render(<ChatInput {...defaultProps} />);
    expect(
      screen.getByPlaceholderText(
        'Click to send. Shift + Enter for a new line.'
      )
    ).toBeInTheDocument();
    expect(screen.getByLabelText('Capture Screenshot')).toBeInTheDocument();
    expect(screen.getByLabelText('Upload File')).toBeInTheDocument();
  });

  it('closes config panel when user starts typing', () => {
    render(<ChatInput {...defaultProps} isConfigPanelVisible={true} />);
    const input = screen.getByPlaceholderText(
      'Click to send. Shift + Enter for a new line.'
    );
    fireEvent.change(input, { target: { value: 'T' } });
    expect(mockSetIsConfigPanelVisible).toHaveBeenCalledWith(false);
  });

  it('handles text input correctly', () => {
    render(<ChatInput {...defaultProps} />);
    const input = screen.getByPlaceholderText(
      'Click to send. Shift + Enter for a new line.'
    );
    fireEvent.change(input, { target: { value: 'Test message' } });
    expect(input).toHaveValue('Test message');
  });

  it('submits the form when Enter is pressed', () => {
    render(<ChatInput {...defaultProps} />);
    const input = screen.getByPlaceholderText(
      'Click to send. Shift + Enter for a new line.'
    );
    fireEvent.change(input, { target: { value: 'Test message' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });
    expect(mockOnSubmit).toHaveBeenCalledWith('Test message', []);
  });

  it('adds a new line when Shift+Enter is pressed', () => {
    render(<ChatInput {...defaultProps} />);
    const input = screen.getByPlaceholderText(
      'Click to send. Shift + Enter for a new line.'
    );
    fireEvent.change(input, { target: { value: 'Line 1' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter', shiftKey: true });
    expect(input).toHaveValue('Line 1\n');
  });

  it('adjusts textarea height as user types', () => {
    render(<ChatInput {...defaultProps} />);
    const textarea = screen.getByPlaceholderText(
      'Click to send. Shift + Enter for a new line.'
    );
    const initialHeight = textarea.style.height;
    fireEvent.change(textarea, {
      target: { value: 'Line 1\nLine 2\nLine 3\nLine 4' }
    });

    expect(textarea.style.height).not.toBe(initialHeight);
  });

  it('enables submit button when there is input', () => {
    render(<ChatInput {...defaultProps} />);
    const input = screen.getByPlaceholderText(
      'Click to send. Shift + Enter for a new line.'
    );
    const submitButton = screen
      .getAllByRole('button')
      .find(button => (button as HTMLButtonElement).type === 'submit');
    fireEvent.change(input, { target: { value: 'Test message' } });
    expect(submitButton).not.toBeDisabled();
  });

  it('disables screen capture feature when isVisionModel is false', () => {
    render(<ChatInput {...defaultProps} isVisionModel={false} />);
    expect(screen.getByLabelText('Capture Screenshot')).toBeDisabled();
  });

  it('handles file upload correctly', async () => {
    const user = userEvent.setup();
    render(<ChatInput {...defaultProps} />);

    const file = new File(['dummy content'], 'test.png', { type: 'image/png' });
    const input = screen.getByLabelText('Upload file', {
      selector: 'input[type="file"]'
    }) as HTMLInputElement;

    await user.upload(input, file);

    expect(input.files?.[0]).toBe(file);
    expect(input.files?.item(0)).toBe(file);
    expect(input.files).toHaveLength(1);
    expect(
      await screen.findByRole('button', {
        name: /Click to view larger file 1/i
      })
    ).toBeInTheDocument();
  });

  it('handles multiple files uploads', async () => {
    const user = userEvent.setup();

    render(<ChatInput {...defaultProps} />);

    const file1 = new File(['dummy content 1'], 'test1.png', {
      type: 'image/png'
    });
    const file2 = new File(['dummy content 2'], 'test2.png', {
      type: 'image/png'
    });
    const input = screen.getByLabelText('Upload file', {
      selector: 'input[type="file"]'
    });

    await user.upload(input, file1);
    await user.upload(input, file2);

    expect(
      await screen.findByRole('button', {
        name: /Click to view larger file 1/i
      })
    ).toBeInTheDocument();
    expect(
      await screen.findByRole('button', {
        name: /Click to view larger file 2/i
      })
    ).toBeInTheDocument();
  });

  it('handles file upload when isVisionModel is false', async () => {
    const user = userEvent.setup();
    render(<ChatInput {...defaultProps} isVisionModel={false} />);

    const file = new File(['dummy content'], 'test.png', { type: 'image/png' });
    const input = screen.getByLabelText('Upload file', {
      selector: 'input[type="file"]'
    }) as HTMLInputElement;

    await user.upload(input, file);

    expect(
      await screen.findByText('The model only supports text messages.')
    ).toBeInTheDocument();
  });

  it('disables file upload when selectedModel.contextWindow is small', () => {
    render(
      <ChatInput
        {...defaultProps}
        selectedModel={{ ...defaultProps.selectedModel, contextWindow: 2000 }}
      />
    );

    const uploadButton = screen.getByLabelText('Upload File');
    expect(uploadButton).toBeDisabled();

    const tooltip = screen.getByText(
      "ai-model's context window is too small to add a file"
    );
    expect(tooltip).toBeInTheDocument();
  });

  it('handles file deletion', async () => {
    const user = userEvent.setup();
    render(<ChatInput {...defaultProps} />);
    const file = new File(['dummy content'], 'test.png', { type: 'image/png' });
    const input = screen.getByLabelText('Upload file', {
      selector: 'input[type="file"]'
    });

    await user.upload(input, file);
    expect(
      await screen.findByRole('button', {
        name: /Click to view larger file 1/i
      })
    ).toBeInTheDocument();

    const deleteButton = screen.getByLabelText('Delete file 1');
    await fireEvent.click(deleteButton);

    expect(
      screen.queryByRole('button', { name: /Click to view larger file 1/i })
    ).not.toBeInTheDocument();
  });

  it('handles screen capture correctly', async () => {
    global.fetch = jest.fn().mockResolvedValueOnce({
      ok: true,
      json: () =>
        Promise.resolve({
          message: 'Success',
          base64Image: 'mocked-base64-string',
          type: 'image/png',
          size: 1000,
          name: 'screenshot.png'
        })
    });

    render(<ChatInput {...defaultProps} />);
    const captureButton = screen.getByLabelText('Capture Screenshot');
    fireEvent.click(captureButton);

    await waitFor(() => {
      const viewImageButton = screen.getByRole('button', {
        name: /Click to view larger file/i
      });
      expect(viewImageButton).toBeInTheDocument();
    });
  });

  it('handles screen capture when isVisionModel is false', async () => {
    render(<ChatInput {...defaultProps} isVisionModel={false} />);
    const captureButton = screen.getByLabelText('Capture Screenshot');

    expect(captureButton).toBeDisabled();
  });

  it('handles screen capture error', async () => {
    global.fetch = jest.fn().mockResolvedValueOnce({
      ok: false,
      json: () =>
        Promise.resolve({
          message: 'Failed to capture screenshot'
        })
    });

    render(<ChatInput {...defaultProps} />);
    const captureButton = screen.getByLabelText('Capture Screenshot');
    fireEvent.click(captureButton);

    await waitFor(() => {
      expect(
        screen.getByText('An error occurred while capturing the screen')
      ).toBeInTheDocument();
    });
  });

  it('handles screen capture network error', async () => {
    global.fetch = jest.fn().mockRejectedValueOnce(new Error('Network error'));

    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(<ChatInput {...defaultProps} />);
    const captureButton = screen.getByLabelText('Capture Screenshot');
    fireEvent.click(captureButton);

    await waitFor(() => {
      expect(consoleSpy).toHaveBeenCalled();
      expect(
        screen.getByText('A network error occurred while capturing the screen')
      ).toBeInTheDocument();
    });

    consoleSpy.mockRestore();
  });

  it('handles file upload error for unsupported image format', async () => {
    const { isSupportedImage } = require('@/src/utils/mediaValidationHelper');
    isSupportedImage.mockReturnValueOnce(['Unsupported file format']);

    const user = userEvent.setup();
    render(<ChatInput {...defaultProps} />);

    const file = new File(['dummy content'], 'test.bmp', { type: 'image/bmp' });
    const input = screen.getByLabelText('Upload file', {
      selector: 'input[type="file"]'
    }) as HTMLInputElement;

    await user.upload(input, file);

    expect(
      await screen.findByText('Unsupported file format')
    ).toBeInTheDocument();
  });

  it('sanitizes user input', () => {
    render(<ChatInput {...defaultProps} />);
    const input = screen.getByPlaceholderText(
      'Click to send. Shift + Enter for a new line.'
    );

    fireEvent.change(input, {
      target: { value: '<script>alert("XSS")</script>' }
    });
    expect(sanitizeWithPreserveCode).toHaveBeenCalledWith(
      '<script>alert("XSS")</script>'
    );
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<ChatInput {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot when isVisionModel is false', () => {
    const { asFragment } = render(
      <ChatInput {...defaultProps} isVisionModel={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
