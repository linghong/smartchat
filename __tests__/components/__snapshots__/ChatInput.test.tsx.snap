// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatInput matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col w-full justify-around flex-shrink-0 items-center  my-1"
  >
    <div
      class="flex w-full justify-around items-center border-2 border-indigo-300 bg-indigo-200 bg-opacity-30 rounded-lg"
    >
      <div
        class="flex w-3/12 ms:w-2/12 sm:w-1/12 items-center justify-around md:mx-1"
      >
        <div
          class="relative group"
        >
          <button
            aria-label="Capture Screenshot"
            class="flex items-center justify-center font-bold  rounded cursor-pointer disabled:cursor-not-allowed"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 3H5V5H3V3ZM7 3H9V5H7V3ZM11 3H13V5H11V3ZM15 3H17V5H15V3ZM19 3H21V5H19V3ZM19 7H21V9H19V7ZM3 19H5V21H3V19ZM3 15H5V17H3V15ZM3 11H5V13H3V11ZM3 7H5V9H3V7ZM10.6667 11L11.7031 9.4453C11.8886 9.1671 12.2008 9 12.5352 9H15.4648C15.7992 9 16.1114 9.1671 16.2969 9.4453L17.3333 11H20C20.5523 11 21 11.4477 21 12V20C21 20.5523 20.5523 21 20 21H8C7.44772 21 7 20.5523 7 20V12C7 11.4477 7.44772 11 8 11H10.6667ZM14 18C15.1046 18 16 17.1046 16 16C16 14.8954 15.1046 14 14 14C12.8954 14 12 14.8954 12 16C12 17.1046 12.8954 18 14 18Z"
              />
            </svg>
          </button>
          <span
            class="absolute top-full left-1/2 transform -translate-x-1/2 mt-4 hidden group-hover:block bg-gray-100 text-black text-xs rounded py-1 px-5 whitespace-nowrap"
          >
            Capture Screenshot
          </span>
        </div>
        <div
          class="relative group"
        >
          <button
            aria-label="Upload File"
            class="flex items-center justify-center font-bold  rounded cursor-pointer disabled:cursor-not-allowed"
          >
            <label
              class="flex items-center justify-center  text-white font-bold px-2 rounded cursor-pointer"
              tabindex="0"
            >
              <input
                accept="image/*, .pdf, .html, .xml, .docx, .txt, .yml, .yaml, .tsx, .ts, .js, .cjs, .coffee, .css, .scss, .sass, less, .xsl, .xlsx,  .jsx, .java, .py, .php, .rb, .erb, .rake, .slim, .cpp, .c, .go, .rs, .log, .cjs, .snap, .sh, .md, .csv, .html, .json, .toml, .sh,  .map,  .xml, .lock, .mod, .metal, .sum, .h, .m, .asp, .net, .dll, .cs, .vb, .fs, .config, .resx, .csproj, .sln"
                aria-label="Upload file"
                style="display: none;"
                type="file"
              />
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="26"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(44, 46, 47); background: rgb(250, 250, 251); cursor: pointer; border: 1px solid #d7dadf;"
                viewBox="0 0 24 24"
                width="26"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14.8287 7.75737L9.1718 13.4142C8.78127 13.8047 8.78127 14.4379 9.1718 14.8284C9.56232 15.219 10.1955 15.219 10.586 14.8284L16.2429 9.17158C17.4144 8.00001 17.4144 6.10052 16.2429 4.92894C15.0713 3.75737 13.1718 3.75737 12.0002 4.92894L6.34337 10.5858C4.39075 12.5384 4.39075 15.7042 6.34337 17.6569C8.29599 19.6095 11.4618 19.6095 13.4144 17.6569L19.0713 12L20.4855 13.4142L14.8287 19.0711C12.095 21.8047 7.66283 21.8047 4.92916 19.0711C2.19549 16.3374 2.19549 11.9053 4.92916 9.17158L10.586 3.51473C12.5386 1.56211 15.7045 1.56211 17.6571 3.51473C19.6097 5.46735 19.6097 8.63317 17.6571 10.5858L12.0002 16.2427C10.8287 17.4142 8.92916 17.4142 7.75759 16.2427C6.58601 15.0711 6.58601 13.1716 7.75759 12L13.4144 6.34316L14.8287 7.75737Z"
                />
              </svg>
            </label>
          </button>
          <span
            class="absolute top-full left-1/2 transform -translate-x-1/2 mt-4 hidden group-hover:block bg-gray-100 text-black text-xs rounded py-1 px-5 whitespace-nowrap"
          >
            Upload File
          </span>
        </div>
      </div>
      <form
        class="flex items-center w-8/12 ms:w-9/12 xs:w-10/12 xl:w-11/12 px-2 py-1 rounded-lg hover:bg-indigo-100 focus:outline-none focus:ring focus:ring-stone-300 focus:ring-offset-red"
      >
        <textarea
          aria-label="Enter your message here"
          class="w-full max-h-96 placeholder-gray-400 overflow-y-auto p-3 focus:ring-stone-100 focus:outline-none"
          id="userInput"
          name="userInput"
          placeholder="Click to send. Shift + Enter for a new line."
          rows="1"
          spellcheck="true"
        />
        <button
          aria-label="Send message"
          class="p-1 border-none opacity-60 hover:stone-600 disabled:cursor-not-allowed"
          disabled=""
          type="submit"
        >
          <svg
            class="w-5 h-5 rotate-90 fill-current"
            viewBox="0 0 18 18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.894 2.553a1 1 0 00-1.788 0l-7 14a1 1 0 001.169 1.409l5-1.429A1 1 0 009 15.571V11a1 1 0 112 0v4.571a1 1 0 00.725.962l5 1.428a1 1 0 001.17-1.408l-7-14z"
            >
              <title>
                Arrow Submit Button
              </title>
            </path>
          </svg>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChatInput matches snapshot when isVisionModel is false 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col w-full justify-around flex-shrink-0 items-center  my-1"
  >
    <div
      class="flex w-full justify-around items-center border-2 border-indigo-300 bg-indigo-200 bg-opacity-30 rounded-lg"
    >
      <div
        class="flex w-3/12 ms:w-2/12 sm:w-1/12 items-center justify-around md:mx-1"
      >
        <div
          class="relative group"
        >
          <button
            aria-label="Capture Screenshot"
            class="flex items-center justify-center font-bold  rounded cursor-pointer disabled:cursor-not-allowed"
            disabled=""
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 3H5V5H3V3ZM7 3H9V5H7V3ZM11 3H13V5H11V3ZM15 3H17V5H15V3ZM19 3H21V5H19V3ZM19 7H21V9H19V7ZM3 19H5V21H3V19ZM3 15H5V17H3V15ZM3 11H5V13H3V11ZM3 7H5V9H3V7ZM10.6667 11L11.7031 9.4453C11.8886 9.1671 12.2008 9 12.5352 9H15.4648C15.7992 9 16.1114 9.1671 16.2969 9.4453L17.3333 11H20C20.5523 11 21 11.4477 21 12V20C21 20.5523 20.5523 21 20 21H8C7.44772 21 7 20.5523 7 20V12C7 11.4477 7.44772 11 8 11H10.6667ZM14 18C15.1046 18 16 17.1046 16 16C16 14.8954 15.1046 14 14 14C12.8954 14 12 14.8954 12 16C12 17.1046 12.8954 18 14 18Z"
              />
            </svg>
          </button>
          <span
            class="absolute top-full left-1/2 transform -translate-x-1/2 mt-4 hidden group-hover:block bg-gray-100 text-black text-xs rounded py-1 px-5 whitespace-nowrap"
          >
            gemini-1.5-pro-exp-0827 does not have vision feature
          </span>
        </div>
        <div
          class="relative group"
        >
          <button
            aria-label="Upload File"
            class="flex items-center justify-center font-bold  rounded cursor-pointer disabled:cursor-not-allowed"
          >
            <label
              class="flex items-center justify-center  text-white font-bold px-2 rounded cursor-pointer"
              tabindex="0"
            >
              <input
                accept="image/*, .pdf, .html, .xml, .docx, .txt, .yml, .yaml, .tsx, .ts, .js, .cjs, .coffee, .css, .scss, .sass, less, .xsl, .xlsx,  .jsx, .java, .py, .php, .rb, .erb, .rake, .slim, .cpp, .c, .go, .rs, .log, .cjs, .snap, .sh, .md, .csv, .html, .json, .toml, .sh,  .map,  .xml, .lock, .mod, .metal, .sum, .h, .m, .asp, .net, .dll, .cs, .vb, .fs, .config, .resx, .csproj, .sln"
                aria-label="Upload file"
                style="display: none;"
                type="file"
              />
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="26"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(44, 46, 47); background: rgb(250, 250, 251); cursor: pointer; border: 1px solid #d7dadf;"
                viewBox="0 0 24 24"
                width="26"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14.8287 7.75737L9.1718 13.4142C8.78127 13.8047 8.78127 14.4379 9.1718 14.8284C9.56232 15.219 10.1955 15.219 10.586 14.8284L16.2429 9.17158C17.4144 8.00001 17.4144 6.10052 16.2429 4.92894C15.0713 3.75737 13.1718 3.75737 12.0002 4.92894L6.34337 10.5858C4.39075 12.5384 4.39075 15.7042 6.34337 17.6569C8.29599 19.6095 11.4618 19.6095 13.4144 17.6569L19.0713 12L20.4855 13.4142L14.8287 19.0711C12.095 21.8047 7.66283 21.8047 4.92916 19.0711C2.19549 16.3374 2.19549 11.9053 4.92916 9.17158L10.586 3.51473C12.5386 1.56211 15.7045 1.56211 17.6571 3.51473C19.6097 5.46735 19.6097 8.63317 17.6571 10.5858L12.0002 16.2427C10.8287 17.4142 8.92916 17.4142 7.75759 16.2427C6.58601 15.0711 6.58601 13.1716 7.75759 12L13.4144 6.34316L14.8287 7.75737Z"
                />
              </svg>
            </label>
          </button>
          <span
            class="absolute top-full left-1/2 transform -translate-x-1/2 mt-4 hidden group-hover:block bg-gray-100 text-black text-xs rounded py-1 px-5 whitespace-nowrap"
          >
            Upload File
          </span>
        </div>
      </div>
      <form
        class="flex items-center w-8/12 ms:w-9/12 xs:w-10/12 xl:w-11/12 px-2 py-1 rounded-lg hover:bg-indigo-100 focus:outline-none focus:ring focus:ring-stone-300 focus:ring-offset-red"
      >
        <textarea
          aria-label="Enter your message here"
          class="w-full max-h-96 placeholder-gray-400 overflow-y-auto p-3 focus:ring-stone-100 focus:outline-none"
          id="userInput"
          name="userInput"
          placeholder="Click to send. Shift + Enter for a new line."
          rows="1"
          spellcheck="true"
        />
        <button
          aria-label="Send message"
          class="p-1 border-none opacity-60 hover:stone-600 disabled:cursor-not-allowed"
          disabled=""
          type="submit"
        >
          <svg
            class="w-5 h-5 rotate-90 fill-current"
            viewBox="0 0 18 18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.894 2.553a1 1 0 00-1.788 0l-7 14a1 1 0 001.169 1.409l5-1.429A1 1 0 009 15.571V11a1 1 0 112 0v4.571a1 1 0 00.725.962l5 1.428a1 1 0 001.17-1.408l-7-14z"
            >
              <title>
                Arrow Submit Button
              </title>
            </path>
          </svg>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
